pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,404 @@
+			UpdateFunc: l7c.updateNamespace,
+		},
+		resyncPeriod,
+	)
+	return l7c
+}
+
+func (l7c *L7VisibilityController) Run(stopCh <-chan struct{}) {
+	defer l7c.queue.ShutDown()
+	klog.InfoS("Starting", "Controller", controllerName)
+	defer klog.InfoS("Shutting down", "Controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, l7c.podListerSynced, l7c.namespaceListerSynced) {
+		return
+	}
+	if itf, ok := l7c.interfaceStore.GetInterfaceByName(config.L7NetworkPolicyTargetPortName); ok {
+		l7c.targetPort = uint32(itf.OFPort)
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(l7c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (l7c *L7VisibilityController) worker() {
+	for l7c.processNextWorkItem() {
+	}
+}
+
+func (l7c *L7VisibilityController) processNextWorkItem() bool {
+	obj, quit := l7c.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		l7c.queue.Forget(key)
+		klog.ErrorS(nil, "Expected string in work queue but got", "key", obj)
+		return true
+	} else if err := l7c.syncPod(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		l7c.queue.Forget(key)
+	} else if err == errInvalidAnnotation || err == errPodInterfaceNotFound {
+		// Handle errors
+		l7c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		l7c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing Pod object for L7Visibility failed, requeue", "Pod", key)
+	}
+	return true
+}
+
+func (l7c *L7VisibilityController) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if !isValidPod(pod) {
+		return
+	}
+	_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		if !l7c.namespaceAnnotationExists(pod) {
+			return
+		}
+	}
+	klog.V(2).InfoS("Processing Pod ADD event", "Pod", klog.KObj(pod))
+	podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+	l7c.queue.Add(podNN)
+}
+
+func (l7c *L7VisibilityController) updatePod(oldObj interface{}, obj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	updatedPod := obj.(*v1.Pod)
+	if !isValidPod(updatedPod) {
+		return
+	}
+	oldAnnotation, _ := oldPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	updatedAnnotation, updatedAnnotationOk := updatedPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	if oldPod.Status.PodIP != updatedPod.Status.PodIP {
+		if !updatedAnnotationOk && !l7c.namespaceAnnotationExists(updatedPod) {
+			return
+		}
+	} else if oldPod.Status.PodIP == updatedPod.Status.PodIP {
+		if oldAnnotation != updatedAnnotation && updatedAnnotationOk {
+			if l7c.namespaceAnnotationExists(updatedPod) {
+				// new need to add to queue, NS event took care of the Pod
+				return
+			}
+		}
+	}
```

Comment:
IMO we should keep this simple and less error prone by enqueuing any Pod for which the PodIP has changed or the annotation has changed. Then we can figure out what to do in the `syncPod` function.

Reply:
I understand your concern and I differ by keeping the queue small and filtering the unconcerned events beforehand. Is there a benefit of not handling it here but in syncPod function ?

Reply:
the number of enqueued events is not as much of a concern here as keeping the code simple and bug free. With the current code, I don't have strong confidence that there is no edge case that we are not covering. Keep in mind that we are only looking at local Pods for that Node, so typically we don't have that many events.

In general, in a controller, it is better to compare the desired state with the actual state of the system, and then decide if any change is required. Obviously, we can still eliminate simple cases early (no change to the IP or annotation).

Reply:
just to confirm should we overwrite a namespace behavior to a pod's behavior, what I mean by this is if a namespace has annotation of any kind, and then a pod gets a new annotation of any kind. we will first remove the pod TCmark flow and provide it with the new TCmarkflow ?

Reply:
New behavior of the code:

Updated Pod:
 if new IP is received and the annotation for L7visibility is present, we will enqueue the Pod and check if the PodNN and direction was already in the PodNNToDirectionMap, return if was.
 
 if no new IP change and also the old and new annotation for L7Visibility are same, we return.
 
 For all other scenarios we enqueue the Pod and process accordingly.